GH-117759: Update GC docs for incremental collection (#126695)
python/cpython: InternalDocs/garbage_collector.md

   the reference counts fall to 0, triggering the destruction of all unreachable
   objects.

Optimization: incremental collection
====================================

In order to limit the time each garbage collection takes, the GC implementation
for the default build uses incremental collection with two generations.

Generational garbage collection takes advantage of what is known as the weak
generational hypothesis: Most objects die young.
This has proven to be very close to the reality of many Python
programs as many temporary objects are created and destroyed very quickly.

To take advantage of this fact, all container objects are segregated into
two generations: young and old. Every new object starts in the young generation.
In order to keep pause times down, scanning of the old generation of the heap
occurs in increments. To keep track of what has been scanned,
the old generation contains two lists: scanned and unscanned heap.

To detect and collect all unreachable objects in the heap, the garbage collector

> must scan the whole heap. This whole heap scan is called a full scavenge.

Reviewer: My understanding is that a full scavenge is a complete sequence of increments (right?) This could be made clearer.

To limit the time each garbage collection takes, the detection and collection
algorithm is executed only on a portion of the heap called an increment.

> For each full scavenge, the increments will cover the whole heap.

Reviewer: This paragraph is somewhat repetitive of lines 366-9.

Each increment, the portion of the heap scanned by a single collection is made up
of three parts:

* The young generation

> * The old generation's least recently objects

Reviewer: ```suggestion
* The old generation's least recently created? used? objects
```

* All objects reachable from those objects that have not yet been scanned this full scavenge

> Any young generation objects surviving this collection are moved to the old generation,

Reviewer: define "surviving".

Author: They aren't collected. It is standard terminology.

and reachable objects in the old generation remain in the old generation.
The old generation is composed of two lists: scanned and unscanned.
(The implementation refers to the unscanned part as `pending` and the scanned part
as `visited`).
Survivors are moved to the back of the scanned list. The old part of increment is taken
from the front of the unscanned list.

When a full scavenge starts, no objects in the heap are considered to have been scanned.
When all objects in the heap have been scanned a cycle ends, and all objects are

> considered unscanned again.

Reviewer: How are they marked "unscanned" in the implementation?

In order to collect all unreachable cycles, each increment must contain all of
an unreachable cycle, or none of it.
In order to make sure that the whole of any unreachable cycle is contained in an
increment,  all unscanned objects reachable from any object in the increment must
be included in the increment.
Thus, to form a complete increment we perform a
[transitive closure](https://en.wikipedia.org/wiki/Transitive_closure)
over reachable, unscanned objects from the initial increment.
We can exclude scanned objects, as they must have been reachable when scanned.
If a scanned object becomes part of an unreachable cycle after being scanned, it

> will not be collected this cycle, but it will be collected next full scavenge.

Reviewer: ```suggestion
will not be collected this cycle, but it will be collected in the next full scavenge.
```

> [!NOTE]
> The GC implementation for the free-threaded build does not use incremental collection.
> Every collection operates on the entire heap.

In order to decide when to run, the collector keeps track of the number of object
allocations and deallocations since the last collection. When the number of
allocations minus the number of deallocations exceeds `threshold0`,
collection starts. `threshold1` determines the fraction of the old
collection that is included in the increment.
The fraction is inversely proportional to `threshold1`,
as historically a larger `threshold1` meant that old generation

> collections were performed less frequency.

Reviewer: ```suggestion
collections were performed less frequently.
```

(The implementation refers to the unscanned part as `pending` and the scanned part
as `visited`).
Survivors are moved to the back of the scanned list. The old part of increment is taken
from the front of the unscanned list.

When a full scavenge starts, no objects in the heap are considered to have been scanned.
When all objects in the heap have been scanned a cycle ends, and all objects are
considered unscanned again.

In order to collect all unreachable cycles, each increment must contain all of
an unreachable cycle, or none of it.
In order to make sure that the whole of any unreachable cycle is contained in an
increment,  all unscanned objects reachable from any object in the increment must
be included in the increment.
Thus, to form a complete increment we perform a
[transitive closure](https://en.wikipedia.org/wiki/Transitive_closure)
over reachable, unscanned objects from the initial increment.
We can exclude scanned objects, as they must have been reachable when scanned.
If a scanned object becomes part of an unreachable cycle after being scanned, it
will not be collected this cycle, but it will be collected next full scavenge.

> .. note::
> 
>    The GC implementation for the free-threaded build does not use incremental collection.
>    Every collection operates on the entire heap.

Reviewer 2: Per GitHub Markdown syntax https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts
```suggestion
> [!NOTE]
> The GC implementation for the free-threaded build does not use incremental collection.
> Every collection operates on the entire heap.
```

   the reference counts fall to 0, triggering the destruction of all unreachable
   objects.

Optimization: incremental collection
====================================

> In order to limit the time each garbage collection takes, the GC implementation

Reviewer: ```suggestion
In order to bound the length of each garbage collection pause, the GC implementation
```

for the default build uses incremental collection with two generations.

Generational garbage collection takes advantage of what is known as the weak
generational hypothesis: Most objects die young.
This has proven to be very close to the reality of many Python
programs as many temporary objects are created and destroyed very quickly.

To take advantage of this fact, all container objects are segregated into

> two generations: young and old. Every new object starts in the young generation.

Reviewer: Maybe mention here that the size of the young generation is more or less controlled but that of the old generation is not.

Increments
----------

Each full scavenge is performed in a series of increments.
For each full scavenge, the combined increments will cover the whole heap.

For each increment, the portion of the heap scanned by a single collection is
made up of three parts:

* The young generation
* The old generation's least recently scanned objects
* All objects reachable from those objects that have not yet been scanned this full scavenge

The surviving objects (those that are not collected) are moved to the back of the
`visited` list in the old generation.

When a full scavenge starts, no objects in the heap are considered to have been scanned,
so all objects in the old generation must be in the `pending` space.
When all objects in the heap have been scanned a cycle ends, and all objects are moved
to the `pending` list again. To avoid having to traverse the entire list, which list is
`pending` and which is `visited` is determined by a field in the `GCState` struct.
The `visited` and `pending` lists can be swapped by toggling this bit.

Correctness
-----------

In order to collect all unreachable cycles, each increment must contain all of

> an unreachable cycle, or none of it.

Reviewer: The sentence above is a bit confusing - I think you mean that in order for a cycle to be collected, it must be contained in a single increment (not that nothing will be collected unless..)

In order to make sure that the whole of any unreachable cycle is contained in an

> increment,  all unscanned objects reachable from any object in the increment must

Reviewer: I'm finding this a bit confusing - when we do an increment, we start at a random unscathed object (right?) and the find everything reachable from it which is not yet scanned, and then how do we know that this is an unreachable cycle?

Author: We don't. But if any object in the increment is part of such a cycle, then the whole cycle must be part of the increment.
Finding the cycles is the job of the base algorithm, we just need to be sure that we don't include part of a cycle in an increment.

   the reference counts fall to 0, triggering the destruction of all unreachable
   objects.

Optimization: incremental collection
====================================

In order to bound the length of each garbage collection pause, the GC implementation
for the default build uses incremental collection with two generations.

Generational garbage collection takes advantage of what is known as the weak
generational hypothesis: Most objects die young.
This has proven to be very close to the reality of many Python
programs as many temporary objects are created and destroyed very quickly.

To take advantage of this fact, all container objects are segregated into
two generations: young and old. Every new object starts in the young generation.
Each garbage collection scans the entire young generation and part of the old generation.

The time taken to scan the young generation can be controlled by controlling the

> size of the young, but the size of the old generation cannot be controlled.

Reviewer: ```suggestion
The time taken to scan the young generation can be controlled by controlling its
size, but the size of the old generation cannot be controlled.
```

In order to keep pause times down, scanning of the old generation of the heap
occurs in increments.

To keep track of what has been scanned, the old generation contains two lists:

* Those objects that have not yet been scanned, referred to as the `pending` list.
* Those objects that have been scanned, referred to as the `visited` list.

To detect and collect all unreachable objects in the heap, the garbage collector
must scan the whole heap. This whole heap scan is called a full scavenge.

Increments
----------

Each full scavenge is performed in a series of increments.
For each full scavenge, the combined increments will cover the whole heap.

> For each increment, the portion of the heap scanned by a single collection is

Reviewer: what does "by a single collection" mean here?

    ...     pass
    ...

    # Move everything to the old generation so it's easier to inspect

>     # the young generations.

Reviewer: ```suggestion
    # the young generation.
```
